azure/scope/machinepool.go

```diff
@@ -116,9 +126,59 @@ func NewMachinePoolScope(params MachinePoolScopeParams) (*MachinePoolScope, erro
 	}, nil
 }
 
+// InitMachinePoolCache sets cached information about the machine pool to be used in the scope.
+func (m *MachinePoolScope) InitMachinePoolCache(ctx context.Context) error {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "scope.MachinePoolScope.InitMachinePoolCache")
+	defer done()
+
+	if m.cache == nil {
+		var err error
+		m.cache = &MachinePoolCache{}
+
+		m.cache.BootstrapData, err = m.GetBootstrapData(ctx)
+		if err != nil {
+			return err
+		}
+
+		m.cache.VMImage, err = m.GetVMImage(ctx)
+		if err != nil {
+			return err
+		}
+
+		skuCache, err := resourceskus.GetCache(m, m.Location())
+		if err != nil {
+			return err
+		}
+
+		m.cache.VMSKU, err = skuCache.Get(ctx, m.AzureMachinePool.Spec.Template.VMSize, resourceskus.VirtualMachines)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get VM SKU %s in compute api", m.AzureMachinePool.Spec.Template.VMSize)
+		}
+	}
+
+	return nil
+}
+
 // ScaleSetSpec returns the scale set spec.
-func (m *MachinePoolScope) ScaleSetSpec() azure.ScaleSetSpec {
-	return azure.ScaleSetSpec{
+func (m *MachinePoolScope) ScaleSetSpec(ctx context.Context) azure.ResourceSpecGetter {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "scope.MachinePoolScope.ScaleSetSpec")
+	defer done()
+
+	shouldPatchCustomData := false
+	if m.HasReplicasExternallyManaged(ctx) {
+		var err error
+		shouldPatchCustomData, err = m.HasBootstrapDataChanges(ctx)
+		if err != nil {
+			log.Error(err, "failed to check for bootstrap data changes")
```

Comment:
we shouldn't be dropping an error without returning it

Looking at this more closely, it seems that we could be making use of the cache in https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/machinepool.go#L587 instead of getting it again to save us from another API call.

Reply:
Is it a convention for the Spec getter functions like `ScaleSetSpec()` to never return an error? I noticed that was the case in all of the other services, and we were calling the functions that might return an error in the reconciler and then setting the field in the spec.

Reply:
> or just to change the nested call like this

The latter

The main reason for not ever returning an error from a spec getter function and pushing that error down to the reconcile function is that the spec getter is also called by the `Delete()` functions and we shouldn't fail delete base on some k8s client call failing that is only used for create (in this case it's true for bootstrap data) since that might cause a broken cluster/machine/machinepool to be stuck forever and not be able to delete.

Reply:
Sounds good. I just refactored those changes into `InitMachinePoolCache`.